File: ielts-vocab-master/Python/Flask/auth/forgot_password.py
import hashlib
import re
import secrets
from datetime import datetime, timedelta, timezone

from flask import Blueprint, jsonify, request

from db import get_db
from mailer import get_mail_config, smtp_send_mail

forgot_password_bp = Blueprint('forgot_password', __name__)


@forgot_password_bp.route('/forgot_password', methods=['POST'])
def forgot_password():
    body  = request.get_json(silent=True) or {}
    email = body.get('email', '').strip().lower()

    if not re.match(r'^[^@\s]+@[^@\s]+\.[^@\s]+$', email):
        return jsonify({'error': 'Please enter a valid email address.'}), 400

    conn = get_db()
    user = conn.execute('SELECT id, name FROM users WHERE email = ?', (email,)).fetchone()

    if user:
        token      = secrets.token_hex(32)
        token_hash = hashlib.sha256(token.encode()).hexdigest()
        expires    = (datetime.now(timezone.utc) + timedelta(minutes=30)).strftime('%Y-%m-%d %H:%M:%S')

        conn.execute(
            'UPDATE users SET reset_token = ?, reset_token_expires = ? WHERE id = ?',
            (token_hash, expires, user['id']),
        )
        conn.commit()

        config   = get_mail_config() or {}
        base_url = config.get('app_base_url', 'http://localhost:8000')
        link     = base_url.rstrip('/') + '/auth/reset_password.html?token=' + token
        name     = user['name'] or 'there'

        body_text = (
            f'Hi {name},\n\n'
            'Someone requested a password reset for your IELTS Vocab Master account.\n\n'
            'Click the link below to set a new password. This link expires in 30 minutes.\n\n'
            f'{link}\n\n'
            "If you didn't request this, you can safely ignore this email.\n"
        )
        smtp_send_mail(email, 'Reset your IELTS Vocab Master password', body_text)

    # Always respond the same way whether or not the email exists, to avoid leaking account existence.
    return jsonify({'ok': True})

File: IELTS-Vocab/PHP/auth/forgot_password.php
<?php
header('Content-Type: application/json');
require __DIR__ . '/db.php';
require __DIR__ . '/mailer.php';

$body = json_decode(file_get_contents('php://input'), true);
$email = isset($body['email']) ? trim(strtolower($body['email'])) : '';

if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
    http_response_code(400);
    echo json_encode(['error' => 'Please enter a valid email address.']);
    exit;
}

$pdo = get_db();
$stmt = $pdo->prepare('SELECT id, name FROM users WHERE email = ?');
$stmt->execute([$email]);
$user = $stmt->fetch(PDO::FETCH_ASSOC);

if ($user) {
    $token = bin2hex(random_bytes(32));
    $tokenHash = hash('sha256', $token);
    $expires = date('Y-m-d H:i:s', time() + 1800);

    $pdo->prepare('UPDATE users SET reset_token = ?, reset_token_expires = ? WHERE id = ?')
        ->execute([$tokenHash, $expires, $user['id']]);

    $config = get_mail_config();
    $baseUrl = $config['app_base_url'] ?? 'http://localhost:8000';
    $link = rtrim($baseUrl, '/') . '/auth/reset_password.html?token=' . $token;
    $name = $user['name'] ?: 'there';

    $bodyText = "Hi {$name},\n\n"
        . "Someone requested a password reset for your IELTS Vocab Master account.\n\n"
        . "Click the link below to set a new password. This link expires in 30 minutes.\n\n"
        . "{$link}\n\n"
        . "If you didn't request this, you can safely ignore this email.\n";

    smtp_send_mail($email, 'Reset your IELTS Vocab Master password', $bodyText);
}

// Always respond the same way whether or not the email exists, to avoid leaking account existence.
echo json_encode(['ok' => true]);
